### PR TITLE
[Bug] [lang] "ext_arr.shape" should return "ti.Expr" instead of "ti.core.Expr"

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
To reproduce:
```py
import taichi as ti
import numpy as np

ti.init(print_preprocessed=True)

@ti.kernel
def func(b: ti.ext_arr()):
    b.shape[0] + 4

func(np.array([1, 2]))
```
```py
[Taichi] mode=release
[Taichi] version 0.6.39, llvm 10.0.0, commit 12e02bec, win, python 3.8.5
[Taichi] Starting on arch=x64
[Taichi] materializing...
Initial AST:
def func(b: ti.ext_arr()):
    b.shape[0] + 4

Preprocessed:
def func():
    import taichi as ti
    b = ti.decl_ext_arr_arg(ti.i32, 1)
    ti.subscript(b.shape, 0) + 4

Checked:
def func():
    import taichi as ti
    b = ti.decl_ext_arr_arg(ti.i32, 1)
    ti.subscript(b.shape, 0) + 4

Final AST:
def func():
    import taichi as ti
    b = ti.decl_ext_arr_arg(ti.i32, 1)
    ti.subscript(b.shape, 0) + 4

Traceback (most recent call last):
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\Administrator\.vscode\extensions\ms-python.python-2020.9.114305\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
    cli.main()
  File "c:\Users\Administrator\.vscode\extensions\ms-python.python-2020.9.114305\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 430, in main
    run()
  File "c:\Users\Administrator\.vscode\extensions\ms-python.python-2020.9.114305\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 267, in run_file
    runpy.run_path(options.target, run_name=compat.force_str("__main__"))
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\Administrator\taichi_three\a.py", line 10, in <module>
    func(np.array([1, 2]))
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\site-packages\taichi\lang\kernel.py", line 574, in wrapped
    return primal(*args, **kwargs)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\site-packages\taichi\lang\kernel.py", line 501, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\site-packages\taichi\lang\kernel.py", line 370, in materialize
    taichi_kernel = taichi_kernel.define(taichi_ast_generator)
  File "C:\Users\Administrator\AppData\Local\Programs\Python\Python38\lib\site-packages\taichi\lang\kernel.py", line 367, in taichi_ast_generator
    compiled()
  File "c:\Users\Administrator\taichi_three\a.py", line 8, in func
    b.shape[0] + 4
TypeError: unsupported operand type(s) for +: 'taichi_core.Expr' and 'int'
```
That's because `taichi_core.Expr`, i.e. `ti.core.Expr`, is a class directly exported from C++, which doesn't have an `__add__` method. We should use the Python wrapped version `ti.Expr` instead.

### CODE REVIEW

File: taichi/lang_util.cpp [hunk 1]
```diff
@@ -393,7 +393,8 @@ class TypePromotionMapping {
       mapping;
   static PrimitiveType::primitive_type to_primitive_type(const DataType d) {
     auto primitive = dynamic_cast<const PrimitiveType *>(d.get_ptr());
-    TI_ASSERT(primitive);
+    TI_ASSERT_INFO(primitive, "Failed to get primitive type! "
+      "Consider adding `ti.init()` to the first line of your program.");
```
Comment (archibate) [inline taichi/lang_util.cpp:397]: @yuanming-hu Could you fix this potentional bug?

**Review by thinking-tower — APPROVED** [review]
LGTM!

**Review by k-ye — APPROVED** [review]
Thanks!

**Review by yuanming-hu — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**